src/rust/engine/src/nodes.rs

```diff
@@ -1034,49 +1076,90 @@ impl Node for NodeKey {
     };
 
     scope_task_workunit_state(Some(workunit_state), async move {
       let context2 = context.clone();
-      let maybe_watch = if let Some(path) = self.fs_subject() {
+      if let Some(path) = self.fs_subject() {
         let abs_path = context.core.build_root.join(path);
-        context
+        let _ = context
           .core
           .watcher
           .watch(abs_path)
           .map_err(|e| Context::mk_error(&format!("{:?}", e)))
-          .await
-      } else {
-        Ok(())
-      };
+          .map_err(|failure| {
+            if let Some(user_facing_name) = user_facing_name {
+              failure.with_pushed_frame(&user_facing_name)
+            } else {
+              failure
+            }
+          })
+          .await?;
+      }
 
-      let mut result = match maybe_watch {
-        Ok(()) => match self {
-          NodeKey::DigestFile(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::DownloadedFile(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::MultiPlatformExecuteProcess(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::ReadLink(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::Scandir(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::Select(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::Snapshot(n) => n.run(context).map_ok(NodeOutput::from).await,
-          NodeKey::Task(n) => n.run(context).map_ok(NodeOutput::from).await,
-        },
-        Err(e) => Err(e),
+      let result = match self {
+        NodeKey::DigestFile(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|v| (NodeOutput::Digest(v), None))
+            .await
+        }
+        NodeKey::DownloadedFile(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|v| (NodeOutput::Snapshot(v), None))
+            .await
+        }
+        NodeKey::MultiPlatformExecuteProcess(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|r| (NodeOutput::ProcessResult(Box::new(r)), None))
+            .await
+        }
+        NodeKey::ReadLink(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|v| (NodeOutput::LinkDest(v), None))
+            .await
+        }
+        NodeKey::Scandir(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|v| (NodeOutput::DirectoryListing(v), None))
+            .await
+        }
+        NodeKey::Select(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|v| (NodeOutput::Value(v), None))
+            .await
+        }
+        NodeKey::Snapshot(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|v| (NodeOutput::Snapshot(v), None))
+            .await
+        }
+        NodeKey::Task(n) => {
+          n.run_wrapped_node(context)
+            .map_ok(|(v, wk)| (NodeOutput::Value(v), wk))
+            .await
+        }
+      };
+      let new_level = match result {
```

Comment:
Rather than having the result become compound in all cases, you could make the level a local mutable variable, and then move the workunit-changing logic into the `let result =` match statement, like:
```
let mut level = metadata.level;
...

  NodeKey::Task(n) => {
    n.run_wrapped_node(context)
      .map_ok(|(v, wk)| {
        level = wk;
        NodeOutput::Value(v)
       })
      .await
  }
```